include/zephyr/sys/async.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Utilities supporting asynchronous calls
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous function completion.
+ *
+ * @note All callback helpers here are safe to call from an ISR.
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ASYNC_H_
+#define ZEPHYR_INCLUDE_SYS_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup async_utility_apis Async Callback Utilities
+ * @{
+ */
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous completion callbacks
+ * and implemented for some commonly used IPC notification mechanisms.
+ *
+ * @note Implementations of this callback should be ISR ok.
+ *
+ * @param context Optional context from which the callback originated;
+ *                commonly a pointer to a struct device.
+ * @param result The result of the asynchronous call.
+ * @param data Data to be given back to the callback.
+ */
+typedef void (*async_callback_t)(void *context, int result, void *data);
+
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Result of async call */
+	int result;
+
+	/** Semaphore that will be given on completion */
+	struct k_sem *sem;
+};
```

Comment:
It could further reduce application level boilerplate code if we added initializer macros for all the `struct async_*` types.

Reply:
Added _INITIALIZER macros